### ISSUE TITLE
Add `DebugSessionParamsDataKind` which will allow finer granulation of tests.

### ISSUE THREAD

**Issue by kpodsiad** [build-server-protocol/build-server-protocol#249]
[Debug request](https://build-server-protocol.github.io/docs/specification.html#debug-request) has field `data` which content depends on `dataKind`. Currently, there is only one possibility to run/debug tests and it is `SCALA_TEST_SUITES` kind.
https://github.com/build-server-protocol/build-server-protocol/blob/2db6bd5093052058368c5390cfbbdcad4095bb4b/bsp4j/src/main/java/ch/epfl/scala/bsp4j/DebugSessionParamsDataKind.java#L5

However, for this kind, the `data` field is an array of strings, for example `["a.b.myTestSuite, "a.b.c.AnotherTestSuite"]`, which is a very coarse granulation. A single test suite could consist of multiple test cases and it is impossible to represent a test run that consists of single or multiple (but not all) test cases from a particular test suite.

That's why I'd like to propose another `DebugSessionParamsDataKind` - (it's just a working name) `SCALA_TEST_CASES`. `data` for this kind can look like `TestCases[]`, where `TestCases` is defined as
```typescript
interface TestCases {
  suiteName: string,
  tests: string[]
}
```
Assuming that `["a.b.myTestSuite, "a.b.c.AnotherTestSuite"]` are test classes in project, such data structure will allow to represent:
- run for single test within single test suite- `[{suiteName: a.b.myTestSuite, tests: ["test 1"]}]`
- run for multiple test within single test suite - `[{suiteName: a.b.myTestSuite, tests: ["test 1", "test 3"]}]` // "test 2" is omitted
- run for multiple tests within multiple test suites - `[{suiteName: a.b.myTestSuite, tests: ["test 1"]}, {suiteName: a.b.c.AnotherTestSuite, tests: ["test 15"]}]`

`suiteName` together with the values of `tests` can be used to construct [test selector](https://github.com/sbt/test-interface/blob/master/src/main/java/sbt/testing/TestSelector.java) which can be further used to filter out unwanted test cases.

This proposition is based on my experience gained in https://github.com/scalameta/metals/pull/3277. Over the next days/weeks I'll try to do some more research or maybe even small proof of concept that will confirm if the proposed interface is sufficient, but I think it's a good time to open the issue and gather others' feedback about this.

**Comment by adpi2** [comment]
That would be great.

Also if we change the format of the debug test request we should also add the jvm options (and environment variable) , as was suggested [here](https://github.com/scalameta/metals-feature-requests/issues/239) and [here](https://github.com/build-server-protocol/build-server-protocol/issues/143).

A possible format would be:
```ts
export interface ScalaTestSelection {
  /** The build target that contains the test classes. */
  target: BuildTargetIdentifier;

  /** The jvm options for the application. */
  jvmOptions: String[];

  /** The environment variables for the application. */
  environmentVariables?: String[];

  /** The fully qualified names of the test classes in this target and the tests in this test classes */
  classes: ScalaTestSuiteSelection[];
}
```

where `ScalaTestSuiteSelection` could be:
```ts
export interface ScalaTestSuiteSelection {
  /** The test class to run. */
  class: String;

  /** The selected tests to run. */
  tests: String[];
}
```

Also I think we should have a convention that `tests= []` means all the tests.

**Comment by kpodsiad** [comment]
@jastice Does adding a new `DebugSessionParamsDataKind` require adding a new capability for it?

**Comment by adpi2** [comment]
@kpodsiad this new `datakind` will be used in the `debugSession/start` request which is only implemented in Metals for the client-side and the scala-debug-adapter for the server side. And Metals can force the version of the scala-debug-adapter, more or less. So I don't think a new capability is useful.

**Comment by tgodzik** [comment]
Wouldn't we need it for Bloop and sbt though? They need to be able to run a single test suite and startDebugSession is implemented separately.

**Comment by kpodsiad** [comment]
I beg to differ @adpi2.
This new datakind has to be implemented in every build server. [Debug server](https://github.com/scalacenter/scala-debug-adapter/blob/384562f0e83b493c12dbdbd9c53b05a5cec29410/core/src/main/scala/ch/epfl/scala/debugadapter/DebugServer.scala) is started by the build server and it expects among others [DebugeeRunner](https://github.com/scalacenter/scala-debug-adapter/blob/384562f0e83b493c12dbdbd9c53b05a5cec29410/core/src/main/scala/ch/epfl/scala/debugadapter/DebuggeeRunner.scala). 
`scala-debug-adapter` knows nothing about available datakinds, it's completely up to the build server to handle them correctly, creating a `DebugeeRunner` instance from them. Hence adding capability makes perfect sense - the build tool doesn't have to implement this new datakind and the client should be able to handle that fact.

**Comment by adpi2** [comment]
Yes but the sbt implementation is in the sbt-debug-adapter plugin (from scala-debug-adapter repository) whose version is set by Metals, and Metals can also force Bloop version. So all in all I think we can implement this new data kind in the scala-debug-adapter repository, then in Bloop, before we release a new version of Metals.

That being said, I would be okay if you add a capability.

**Comment by adpi2** [comment]
Well it may not be so easy to implement test selection in Bloop and sbt. So I am probably wrong when saying that we can implement it all before we release a new version of Metals.

**Comment by kpodsiad** [comment]
> So all in all I think we can implement this new data kind in the scala-debug-adapter repository, then in Bloop, before we release a new version of Metals.

It's possible and this is exactly what I want to do. However, I want to add capability because of different reason.

We shouldn't forget about other build servers like Mill for example. When someone wants to use Mill as BSP then Metals should know about supported features. This is especially important in this particular datakind -- if build server supports `scala-test-selection` then Metals can e.g. generate code lenses per every test within the test suite.

**Comment by kpodsiad** [comment]
closed via https://github.com/build-server-protocol/build-server-protocol/pull/304
